### PR TITLE
fix multi device dependency issue

### PR DESCRIPTION
**Pull request by panyx0718**
test=develop

### CODE REVIEW

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc [hunk 1]
```diff
@@ -265,6 +265,16 @@ std::vector<ir::Node *> SortOpsAndDelayOptimizeOp(const ir::Graph &graph) {
       std::unordered_set<ir::Node *> optimize_set(optimize_ops.begin(),
                                                   optimize_ops.end());
       for (ir::Node *n : sorted_ret) {
+        // TODO(panyx0718): Revisit this soon.
+        if (boost::get<int>(
+                n->Op()->GetAttr(OpProtoAndCheckerMaker::OpRoleAttrName())) !=
+                static_cast<int>(OpRole::kBackward) &&
+            boost::get<int>(
```
Comment (typhoonzero) [inline paddle/fluid/framework/details/multi_devices_graph_pass.cc]: scale loss op should be skipped too? they are of role `kBackward | kLoss` or `kForward | kLoss`

File: python/paddle/fluid/framework.py [hunk 2]
```diff
@@ -1496,25 +1496,32 @@ def _optimized_guard(self, param_and_grads):
...
         ]
         yield
-        self._op_role_var = []
-        self._current_role = OpRole.Forward
+        self._op_role_var = tmp_var
+        self._current_role = tmp_role
 
     @contextlib.contextmanager
-    def _lr_schedule_guard(self):
+    def _lr_schedule_guard(self, is_with_opt=False):
```
Comment (typhoonzero) [inline python/paddle/fluid/framework.py:1513]: lr schedule ops can always be marked as `optimize | lrsched` I think, the learning rate should only be used by optimize ops.

Comment (chengduoZH) [reply python/paddle/fluid/framework.py:1513]: That is right!

**Review by typhoonzero — APPROVED** [review]

File: python/paddle/fluid/framework.py (hunk 2 above)
Comment (typhoonzero) [reply python/paddle/fluid/framework.py:1513]: Thanks, @panyx0718 explained why should do this.

when memory optimize is enabled, lr sched ops created by `learning_rate_scheduler.py` may depended by backward ops due to variable reusing. Other lr ops defined by user should not have this issue.
